docs/design_notes/module_lifecycle.md

```diff
@@ -0,0 +1,289 @@
+# The Module Lifecycle
+
+This design note is intended for people who are somewhat familiar with Modules but want to understand more about the design principles behind the abstraction. After reading this, you should have a solid understanding of the assumptions and guarantees the Module API is built upon. If you have no practical experience with Modules yet, Please take a look at the [MNIST tutorial](https://flax.readthedocs.io/en/latest/notebooks/annotated_mnist.html).
```

Comment:
What about linking `Module` to `linen.Module` ?
(First, it's nice to have a code link ready when reading the tutorial, second I like having pages that anybody can land on and understand what is meant)

Reply:
done